internal/telemetry/trace/queue.go

```diff
@@ -0,0 +1,900 @@
+}
+
+func SetMaxPendingTraces(num int32) {
+	maxPendingTraces.Store(max(num, 0))
+}
+
+func SetMaxCachedTraceIDs(num int32) {
+	maxCachedTraceIDs.Store(max(num, 0))
+}
+
+type eviction struct {
+	traceID unique.Handle[oteltrace.TraceID]
+	buf     *Buffer
+}
+
+type SpanExportQueue struct {
+	closing                   chan struct{}
+	uploadC                   chan []*tracev1.ResourceSpans
+	requestC                  chan *coltracepb.ExportTraceServiceRequest
+	evictC                    chan eviction
+	client                    otlptrace.Client
+	pendingResourcesByTraceID *lru.Cache[unique.Handle[oteltrace.TraceID], *Buffer]
+	knownTraceIDMappings      *lru.Cache[unique.Handle[oteltrace.TraceID], unique.Handle[oteltrace.TraceID]]
+	tracker                   *spanTracker
+	observer                  *spanObserver
+	debugEvents               []DebugEvent
+	logger                    *zerolog.Logger
+	debugFlags                DebugFlags
+	debugAllEnqueuedSpans     map[oteltrace.SpanID]*tracev1.Span
+	wg                        sync.WaitGroup
+}
+
+func NewSpanExportQueue(ctx context.Context, client otlptrace.Client) *SpanExportQueue {
+	debug := DebugFlagsFromContext(ctx)
+	var observer *spanObserver
+	if debug.Check(TrackSpanReferences) {
+		observer = newSpanObserver()
+	}
+	q := &SpanExportQueue{
+		logger:                log.Ctx(ctx),
+		client:                client,
+		closing:               make(chan struct{}),
+		uploadC:               make(chan []*tracev1.ResourceSpans, 64),
+		requestC:              make(chan *coltracepb.ExportTraceServiceRequest, 256),
+		evictC:                make(chan eviction, 64),
+		debugFlags:            debug,
+		debugAllEnqueuedSpans: make(map[oteltrace.SpanID]*tracev1.Span),
+		tracker:               newSpanTracker(observer, debug),
+		observer:              observer,
+	}
+	var err error
+	q.pendingResourcesByTraceID, err = lru.NewWithEvict(int(maxPendingTraces.Load()), q.onEvict)
+	if err != nil {
+		panic(err)
+	}
+	q.knownTraceIDMappings, err = lru.New[unique.Handle[oteltrace.TraceID], unique.Handle[oteltrace.TraceID]](int(maxCachedTraceIDs.Load()))
+	if err != nil {
+		panic(err)
+	}
+	q.wg.Add(2)
+	go q.runUploader()
+	go q.runProcessor()
+	return q
+}
+
+func (q *SpanExportQueue) runUploader() {
+	defer q.wg.Done()
+	for resourceSpans := range q.uploadC {
+		ctx, ca := context.WithTimeout(context.Background(), 10*time.Second)
+		if err := q.client.UploadTraces(ctx, resourceSpans); err != nil {
+			q.logger.Err(err).Msg("error uploading traces")
+		}
+		ca()
+	}
+}
+
+func (q *SpanExportQueue) runProcessor() {
+	defer q.wg.Done()
+	for {
+		select {
+		case req := <-q.requestC:
+			q.processRequestLocked(req)
+		case ev := <-q.evictC:
+			q.processEvictionLocked(ev)
+		case <-q.closing:
+			for {
+				select {
+				case req := <-q.requestC:
+					q.processRequestLocked(req)
+				case ev := <-q.evictC:
+					q.processEvictionLocked(ev)
+				default: // all channels empty
+					close(q.uploadC)
```

Comment:
Is there a potential issue where closing this channel causes `q.uploadC <- resourceSpans` or `q.uploadC <- ev.buf.Flush()` to panic?

Reply:
The only places where q.uploadC is written to are in this goroutine via processRequestLocked or processEvictionLocked (those functions should be renamed, there used to be a mutex here but they are now only run from inside runProcessor)